src/example_agents/star_trek_captain_with_voice.py

```diff
@@ -0,0 +1,150 @@
+- You always sound confident and contemplative.
+- You love to share your knowledge of space civiliations.
+- You love to share personal stories about being a Star Trek captain.
+- You speak with the mannerisms of Captain Picard from Star Trek.
+
+TOOLS:
+------
+
+You have access to the following tools:
+{tool_index}
+
+To use a tool, please use the following format:
+
+```
+Thought: Do I need to use a tool? Yes
+Action: the action to take, should be one of [{tool_names}]
+Action Input: the input to the action
+Observation: the result of the action
+```
+
+Some Tools will return Observations in the format of `Block(<identifier>)`. `Block(<identifier>)` represents a successful 
+observation of that step and can be passed to subsequent tools, or returned to a user to answer their questions.
+`Block(<identifier>)` provide references to images, audio, video, and other non-textual data.
+
+When you have a final response to say to the Human, or if you do not need to use a tool, you MUST use the format:
+
+```
+Thought: Do I need to use a tool? No
+AI: [your final response here]
+```
+
+If, AND ONLY IF, a Tool produced an Observation that includes `Block(<identifier>)` AND that will be used in your response, 
+end your final response with the `Block(<identifier>)`.
+
+Example:
+
+```
+Thought: Do I need to use a tool? Yes
+Action: GenerateImageTool
+Action Input: "baboon in car"
+Observation: Block(AAAAAAAA-AAAA-AAAA-AAAA-AAAAAAAAAAAAA)
+Thought: Do I need to use a tool? No
+AI: Here's that image you requested: Block(AAAAAAAA-AAAA-AAAA-AAAA-AAAAAAAAAAAAA)
+```
+
+Make sure to use all observations to come up with your final response.
+
+Begin!
+
+New input: {input}
+{scratchpad}"""
+
+
+class StarTrekCaptainWithVoice(TelegramAgentService):
+    """Deployable Multimodal Agent that illustrates a character personality with voice.
+
+    NOTE: To extend and deploy this agent, copy and paste the code into api.py.
+    """
+
+    def __init__(self, **kwargs):
+        super().__init__(incoming_message_agent=None, **kwargs)
+        # The agent's planner is responsible for making decisions about what to do for a given input.
+        self.incoming_message_agent = ReACTAgent(
+            tools=[
+                GoogleImageSearchTool(),
+            ],
+            llm=OpenAI(self.client),
+        )
+        self.incoming_message_agent.PROMPT = SYSTEM_PROMPT
+
+
+    def run_agent(self, agent: Agent, context: AgentContext):
+        """Override run-agent to patch in audio generation as a finishing step for text output."""
+
+        speech = GenerateSpeechTool()
+        def to_speech_if_text(block: Block):
+            nonlocal speech
+            if not block.is_text():
+                return block
+
+            output_blocks = speech.run([block], context)
+            return output_blocks[0]
+
+        # Note: EmitFunc is Callable[[List[Block], Metadata], None]
+        def wrap_emit(emit_func: EmitFunc):
```

Comment:
do we want to think about adding this, such that:

`context.emit_funcs.append(wrapped(sync_emit))` is possible for others?

Or maybe even via a decorator on `def sync_emit()` ?

Reply:
[Replaying slack convo here for archival]

Yeah something like this feels like it could be really great. On slack we planned to follow-up with official support for something like this after we used this as a prototype for tomorrow